### PR TITLE
Feature: Improved Transaction Types

### PR DESCRIPTION
**Pull request by 0xOmarA**
This PR is a complete refactor to the transaction types detection system based on the new requirements of the wallet.

### CODE REVIEW

File: .github/workflows/release-kotlin.yml [hunk 1]
```diff
@@ -37,7 +37,7 @@ jobs:
           rustup toolchain install nightly
 
           # Update both toolchains
-          rustup update stable
+          rustup update nightly-2023-11-24
```
Comment (0xOmarA) [inline .github/workflows/release-kotlin.yml:40]: Sadly needed to do this due to https://github.com/rust-lang/rust/issues/65991

Comment (mstrug-rdx) [reply .github/workflows/release-kotlin.yml:40]: If we switching back to nightly, then I think it would be good to revert `rust-toolchain.toml` file.

Comment (0xOmarA) [reply .github/workflows/release-kotlin.yml:40]: Sounds good, will also add a comment describing why we needed this for this particular PR and when what features need to be stable before we can switch to stable.

File: crates/radix-engine-toolkit/src/transaction_types/traverser/traverser.rs [hunk 2]
```diff
@@ -0,0 +1,794 @@
...
+        callbacks: &mut [&mut dyn ManifestSummaryCallback],
+        instruction: &InstructionV1,
+        instruction_index: usize,
+    ) {
+        // At the beginning of an instruction, call the on_instruction callback
+        callbacks.iter_mut().for_each(|callback| {
+            callback.on_instruction(instruction, instruction_index)
+        });
+
+        // Notify the callbacks of the created created account proofs
```
Comment (mstrug-rdx) [inline crates/radix-engine-toolkit/src/transaction_types/traverser/traverser.rs:47]: Doubled word in comment.

**Review by mstrug-rdx — APPROVED** [review]

File: crates/radix-engine-toolkit/src/transaction_types/traverser/types/transfer.rs [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+        instruction_index: usize,
+    ) {
+        /* Simple transfer accounting */
+        self.instructions_match_simple_transfer &= (instruction_index == 0
+            && matches!(instruction, InstructionV1::CallMethod { address, method_name, .. } if is_account(address) && ACCOUNT_WITHDRAW_METHODS.contains(method_name)))
+            || (instruction_index == 1
+                && matches!(
+                    instruction,
+                    InstructionV1::TakeFromWorktop { .. }
+                        | InstructionV1::TakeNonFungiblesFromWorktop { .. }
```
[the remark this answers is not in the document]
Comment (0xOmarA) [reply crates/radix-engine-toolkit/src/transaction_types/traverser/types/transfer.rs:59]: Good question! According to the [transaction types rules](https://radixdlt.atlassian.net/wiki/spaces/AT/pages/3065774152/Conforming+Manifest+Types+for+Wallet+Display?focusedCommentId=3115155458#1a)-Simple-Transfer), a `TakeAllFromWorktop` is not permitted in this setting. In fact, it seems like there is a mistake here where we allow for `TakeNonFungiblesFromWorktop` instructions when the spec mentions that we should not.

File: crates/radix-engine-toolkit/src/transaction_types/traverser/traverser.rs (hunk 2 above)
Comment (0xOmarA) [reply crates/radix-engine-toolkit/src/transaction_types/traverser/traverser.rs:47]: Good catch! Will change this.

**Pull request merged**